spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/rdd/CassandraJoinRDD.scala

```diff
@@ -0,0 +1,152 @@
+    case PartitionKeyColumns => tableDef.partitionKey.map(col => col.columnName: NamedColumnRef).toSeq
+    case SomeColumns(cs@_*) => {
+      checkColumnsExistence(cs)
+      checkValidJoin(cs)
+    }
+  }
+
+  protected def checkValidJoin(columns: Seq[NamedColumnRef]): Seq[NamedColumnRef] = {
+    val allColumnNames = tableDef.allColumns.map(_.columnName).toSet
+    val regularColumnNames = tableDef.regularColumns.map(_.columnName).toSet
+    val partitionKeyColumnNames = tableDef.partitionKey.map(_.columnName).toSet
+    val joinColumnNames = columns.map(_.columnName).toSet
+
+    def checkSingleColumn(column: NamedColumnRef) = {
+      if (regularColumnNames.contains(column.columnName))
+        throw new IllegalArgumentException(s"Can't pushdown join on column $column because it is not part of the PRIMARY KEY")
+      column
+    }
+
+    //Make sure we have all of the clustering indexes between the 0th position and the max requested in the join
+    val chosenClusteringColumns = tableDef.clusteringColumns
+      .filter(tableCol => columns.exists(joinCol => tableCol.columnName == joinCol.columnName))
+    if (chosenClusteringColumns.size != 0) {
+      val chosenClusteringIndexes = chosenClusteringColumns.flatMap(col => col.componentIndex)
+      val maxIndex = chosenClusteringIndexes.max
+      val requiredIndexes = 0 to maxIndex
+      val missingIndexes = requiredIndexes.toSet -- chosenClusteringIndexes.toSet
+      if (missingIndexes.size != 0) {
+        val maxCol = tableDef.clusteringColumns
+          .filter(cc => cc.componentIndex.get == maxIndex)
+          .map(_.columnName).mkString(", ")
+        val otherCols = tableDef.clusteringColumns
+          .filter(cc => missingIndexes.contains(cc.componentIndex.get))
+          .map(_.columnName).mkString(", ")
+        throw new IllegalArgumentException(s"Can't pushdown join on column ${maxCol} without also specifying [ ${otherCols} ]")
+      }
+    }
+    val missingPartitionKeys = partitionKeyColumnNames -- joinColumnNames
+    if (missingPartitionKeys.size != 0) {
+      throw new IllegalArgumentException(s"Can't join without the full partition key. Missing: [ ${missingPartitionKeys} ]")
+    }
+    columns.map(checkSingleColumn)
+  }
+
+  val rowWriter = implicitly[RowWriterFactory[O]].rowWriter(
+    tableDef,
+    joinColumnNames.map(_.columnName))
+
+  def on(joinColumns: ColumnSelector): CassandraJoinRDD[O, N] = {
+    new CassandraJoinRDD[O, N](prev, keyspaceName, tableName, connector, columnNames, joinColumns, where, readConf).asInstanceOf[this.type]
+  }
+
+  //We need to make sure we get selectedColumnNames before serialization so that our RowReader is
+  //built
+  private val singleKeyCqlQuery: (String) = {
+    require(tableDef.partitionKey.map(_.columnName).exists(
+      partitionKey => where.predicates.exists(_.contains(partitionKey))) == false,
+      "Partition keys are not allowed in the .where() clause of a Cassandra Join")
```

Comment:
You might use a where predicate parser that Artem wrote some time ago to do it better.